### ISSUE TITLE
Enlarge Batch Size Per GPU

### ISSUE THREAD

**Issue by trapoom555** [trapoom555/MiniCPM-Dense-Retriever#6]
### The result of training with `batch_size_per_gpu = 2` is unstable.

<img width="419" alt="image" src="https://github.com/trapoom555/MiniCPM-Dense-Retriever/assets/36850068/254f97a7-bc7f-4291-960e-5a22f5a06a99">

- The key success of contrastive learning heavily depends on batch size
- `batch_size_per_gpu = 2` has already consumed ~45% of GPU memory (RTX3090)
- Now using a normal DDP strategy + bf16 Mixed Precision + LoRA for training

**Comment by trapoom555** [comment]
### Try pushing the limit of DDP strategy
- Preprocessed data before training (having 150 max length, longer than this will be truncated)
- Train with `batch_size_per_gpu=8` (takes 22GB of GPU RAM, ~92%)

<img width="423" alt="image" src="https://github.com/trapoom555/MiniCPM-Dense-Retriever/assets/36850068/8715f597-e3ce-4f44-b26e-7660b9527add">

- The result is better than `batch_size_per_gpu = 2` experiment. Emphasize that **increasing of batch size leads to better performance in Contrastive Learning**.
- Will try off-load strategy in **DeepSpeed / FSDP** to maximize the number of batch size

**Comment by trapoom555** [comment]
### FSDP with Huggingface PEFT

The code has been migrated from Lightning to Huggingface format and applied FSDP strategy with CPU offload.
`batch_size_per_gpu` can be scaled up to `10` giving much stable performance.

<img width="352" alt="image" src="https://github.com/trapoom555/MiniCPM-Dense-Retriever/assets/36850068/52be84ca-821d-4140-87a6-a06a89fdb39d">

To be concise, `batch_size_per_gpu=10` means that the optimization process is actually using `batch_size=30` because there're anchors, positive samples and negative samples. InfoICE loss uses `19` samples as its negative samples.

Another finding is that using larger gradient accumulation, in this case `gradient_accumulation=3`, will make the result even more stable.

<img width="346" alt="image" src="https://github.com/trapoom555/MiniCPM-Dense-Retriever/assets/36850068/0b103824-a2b1-4103-9629-0053d756749a">

**Comment by trapoom555** [comment]
For **DeepSpeed**, I've tried implementing the code in [deepspeed](https://github.com/trapoom555/MiniCPM-Dense-Retriever/tree/deepspeed) branch but I'm not able to use the offload strategy. It requires external packages to be installed and I don't have sudo permission in the server. This requires too much effort for me to install things.
